File: java/extensions/semantickernel-planners/src/main/java/com/microsoft/semantickernel/planner/actionplanner/ActionPlanner.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.planner.actionplanner;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.microsoft.semantickernel.Kernel;
import com.microsoft.semantickernel.SKBuilders;
import com.microsoft.semantickernel.orchestration.SKContext;
import com.microsoft.semantickernel.orchestration.SKFunction;
import com.microsoft.semantickernel.orchestration.WritableContextVariables;
import com.microsoft.semantickernel.planner.PlanningException;
import com.microsoft.semantickernel.semanticfunctions.PromptTemplateConfig;
import com.microsoft.semantickernel.skilldefinition.ReadOnlySkillCollection;
import com.microsoft.semantickernel.skilldefinition.annotations.DefineSKFunction;
import com.microsoft.semantickernel.skilldefinition.annotations.SKFunctionParameters;
import com.microsoft.semantickernel.textcompletion.CompletionRequestSettings;
import com.microsoft.semantickernel.textcompletion.CompletionSKFunction;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import javax.annotation.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import reactor.core.publisher.Mono;

/// <summary>
/// Action Planner allows to select one function out of many, to achieve a given goal.
/// The planner implement the Intent Detection pattern, uses the functions registered
/// in the kernel to see if there's a relevant one, providing instructions to call the
/// function and the rationale used to select it. The planner can also return
/// "no function" is nothing relevant is available.
/// The rationale is currently available only in the prompt, we might include it in
/// the Plan object in future.
/// </summary>
public class ActionPlanner {
    private static final Logger LOGGER = LoggerFactory.getLogger(ActionPlanner.class);

    // Extracts the json portion of a plan between the braces, stripping off any non-json content
    private static final Pattern CLEAN_PLAN =
            Pattern.compile("[^{]*(\\{.*})[^}]*", Pattern.MULTILINE | Pattern.DOTALL);
    private static final String StopSequence = "#END-OF-PLAN";
    private static final String SkillName = "this";

    // Planner semantic function
    private final CompletionSKFunction plannerFunction;

    // Context used to access the list of functions in the kernel
    private SKContext context;
    private Kernel kernel;

    // TODO: allow to inject skill store

    /**
     * Initialize a new instance of the ActionPlanner class
     *
     * @param kernel The semantic kernel instance
     * @param prompt prompt override
     */
    public ActionPlanner(Kernel kernel, @Nullable String prompt) {
        // Verify.NotNull(kernel);
        String promptTemplate;
        if (prompt == null) {
            promptTemplate = read("skprompt.txt");
        } else {
            promptTemplate = prompt;
        }

        this.plannerFunction =
                SKBuilders.completionFunctions()
                        .withKernel(kernel)
                        .setPromptTemplate(promptTemplate)
                        .setSkillName(SkillName)
                        .setCompletionConfig(
                                new PromptTemplateConfig.CompletionConfig(0.0, 0.0, 0.0, 0.0, 1024))
                        .build();

        kernel.importSkill(this, SkillName);

        this.kernel = kernel;
        this.context = SKBuilders.context().setSkills(kernel.getSkills()).build();
    }

    public static String read(String file) {
        try (InputStream stream = ActionPlanner.class.getResourceAsStream(file)) {
            byte[] buffer = new byte[stream.available()];
            stream.read(buffer);
            return new String(buffer, StandardCharsets.UTF_8);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public Mono<Plan> createPlanAsync(String goal) {
        if (goal == null || goal.isEmpty()) {
            throw new PlanningException(
                    PlanningException.ErrorCodes.InvalidGoal, "The goal specified is empty");
        }

        context = this.context.update(goal);

        return this.plannerFunction
                .invokeAsync(this.context, new CompletionRequestSettings(0, 0, 0, 0, 2048))
                .handle(
                        (result, sink) -> {
                            try {
                                sink.next(parsePlan(goal, result));
                            } catch (Exception e) {
                                sink.error(e);
                            }
                        });
    }

    private Plan parsePlan(String goal, SKContext result) {
        ActionPlanResponse planData;
        // Clean up the plan, removing any prompt and the stop sequence
        Matcher matcher = CLEAN_PLAN.matcher(result.getResult());

        String plan = result.getResult();
        if (matcher.matches()) {
            plan = matcher.group(1);
        }

        try {
            planData = new ObjectMapper().readValue(plan, ActionPlanResponse.class);
        } catch (Exception e) {
            throw new PlanningException(
                    PlanningException.ErrorCodes.InvalidPlan,
                    "Plan parsing error, invalid JSON",
                    e);
        }

        if (planData == null) {
            throw new PlanningException(
                    PlanningException.ErrorCodes.InvalidPlan,
                    "The plan deserialized to a null object");
        }

        SKFunction function = getPlanFunction(planData);

        WritableContextVariables variables = SKBuilders.variables().build().writableClone();

        planData.plan
                .parameters
                .entrySet()
                .forEach(
                        entry -> {
                            variables.setVariable(entry.getKey(), entry.getValue());
                        });

        if (function == null) {
            // No function was found - return a plan with no steps.
            return new Plan(goal, variables, () -> kernel.getSkills());
        } else {
            return new Plan(goal, variables, () -> kernel.getSkills(), function);
        }
    }

    public SKFunction getPlanFunction(ActionPlanResponse planData) {

        if (planData.plan.function.contains(".")) {
            String[] parts = planData.plan.function.split("\\.", -1);
            SKFunction function =
                    context.getSkills().getFunction(parts[0], parts[1], SKFunction.class);

            if (function == null) {
                throw new PlanningException(
                        PlanningException.ErrorCodes.InvalidPlan,
                        "Unknown function " + planData.plan.function);
            }
            return function;

        } else if (!planData.plan.function.isEmpty()) {
            CompletionSKFunction function =
                    context.getSkills()
                            .getFunction(planData.plan.function, CompletionSKFunction.class);

            if (function == null) {
                throw new PlanningException(
                        PlanningException.ErrorCodes.InvalidPlan,
                        "Unknown skill " + planData.plan.function);
            }

            return function;
        }
        return null;
    }

    // TODO: use goal to find relevant functions in a skill store
    /// <summary>
    /// Native function returning a list of all the functions in the current context,
    /// excluding functions in the planner itself.
    /// </summary>
    /// <param name="goal">Currently unused. Will be used to handle long lists of functions.</param>
    /// <param name="context">Function execution context</param>
    /// <returns>List of functions, formatted accordingly to the prompt</returns>

    @DefineSKFunction(
            name = "listOfFunctions",
            description = "List all functions available in the kernel")
    @Nullable
    public String listOfFunctions(
            @SKFunctionParameters(
                            name = "goal",
                            description = "The current goal processed by the planner",
                            defaultValue = "")
                    String goal,
            SKContext context) {
        // Verify.NotNull(context.Skills);
        ReadOnlySkillCollection skills = context.getSkills();

        return this.populateList(skills);
    }

    // TODO: generate string programmatically
    // TODO: use goal to find relevant examples
    @DefineSKFunction(
            name = "GoodExamples",
            description = "List a few good examples of plans to generate")
    public String goodExamples(
            @SKFunctionParameters(
                            name = "goal",
                            description = "The current goal processed by the planner",
                            defaultValue = "")
                    String goal,
            SKContext context) {

        return "\n"
                + "[EXAMPLE]\n"
                + "- List of functions:\n"
                + "// Read a file.\n"
                + "FileIOSkill.ReadAsync\n"
                + "Parameter \"path\": Source file.\n"
                + "// Write a file.\n"
                + "FileIOSkill.WriteAsync\n"
                + "Parameter \"path\": Destination file. (default value: sample.txt)\n"
                + "Parameter \"content\": File content.\n"
                + "// Get the current time.\n"
                + "TimeSkill.Time\n"
                + "No parameters.\n"
                + "// Makes a POST request to a uri.\n"
                + "HttpSkill.PostAsync\n"
                + "Parameter \"body\": The body of the request.\n"
                + "- End list of functions.\n"
                + "Goal: create a file called \"something.txt\".\n"
                + "{\"plan\":{\n"
                + "\"rationale\": \"the list contains a function that allows to create files\",\n"
                + "\"function\": \"FileIOSkill.WriteAsync\",\n"
                + "\"parameters\": {\n"
                + "\"path\": \"something.txt\",\n"
                + "\"content\": null\n"
                + "}}}\n"
                + "#END-OF-PLAN\n";
    }

    @DefineSKFunction(
            name = "EdgeCaseExamples",
            description = "List a few edge case examples of plans to handle")
    public String edgeCaseExamples(
            @SKFunctionParameters(
                            name = "goal",
                            description = "The current goal processed by the planner",
                            defaultValue = "")
                    String goal,
            SKContext context) {
        return "\n"
                + "[EXAMPLE]\n"
                + "- List of functions:\n"
                + "// Get the current time.\n"
                + "TimeSkill.Time\n"
                + "No parameters.\n"
                + "// Write a file.\n"
                + "FileIOSkill.WriteAsync\n"
                + "Parameter \"path\": Destination file. (default value: sample.txt)\n"
                + "Parameter \"content\": File content.\n"
                + "// Makes a POST request to a uri.\n"
                + "HttpSkill.PostAsync\n"
                + "Parameter \"body\": The body of the request.\n"
                + "// Read a file.\n"
                + "FileIOSkill.ReadAsync\n"
                + "Parameter \"path\": Source file.\n"
                + "- End list of functions.\n"
                + "Goal: tell me a joke.\n"
                + "{\"plan\":{\n"
                + "\"rationale\": \"the list does not contain functions to tell jokes or"
                + " something funny\",\n"
                + "\"function\": \"\",\n"
                + "\"parameters\": {\n"
                + "}}}\n"
                + "#END-OF-PLAN\n";
    }

    private String populateList(ReadOnlySkillCollection skills) {
        return skills.getAllFunctions().getAll().stream()
                .filter(skill -> !SkillName.equalsIgnoreCase(skill.getSkillName()))
                .map(
                        func -> {
                            StringBuilder result = new StringBuilder();
                            // Function description
                            if (func.getDescription() != null) {
                                result.append("// " + addPeriod(func.getDescription()) + "\n");
                            } else {
                                LOGGER.warn(
                                        "{0}.{1} is missing a description",
                                        func.getSkillName(), func.getName());
                                result.append(
                                        "// Function "
                                                + func.getSkillName()
                                                + "."
                                                + func.getName()
                                                + ".\n");
                            }

                            // Function name
                            result.append(func.getSkillName() + "." + func.getName() + "\n");

                            func.describe()
                                    .getParameters()
                                    .forEach(
                                            parameter -> {
                                                String description;
                                                if (parameter.getDescription() == null
                                                        || parameter.getDefaultValue().isEmpty()) {
                                                    description = parameter.getName();
                                                } else {
                                                    description = parameter.getDescription();
                                                }

                                                String defaultValueString;
                                                if (parameter.getDefaultValue() == null
                                                        || parameter.getDefaultValue().isEmpty()) {
                                                    defaultValueString = "";
                                                } else {
                                                    defaultValueString =
                                                            " (default value: "
                                                                    + parameter.getDefaultValue()
                                                                    + ")";
                                                }

                                                result.append(
                                                        "Parameter \""
                                                                + parameter.getName()
                                                                + "\": "
                                                                + addPeriod(description)
                                                                + " "
                                                                + defaultValueString
                                                                + "\n");
                                            });

                            return result.toString();
                        })
                .reduce("", (a, b) -> a + b, (a, b) -> a + b);
    }

    private static String addPeriod(String x) {
        return x.endsWith(".") ? x : x + ".";
    }
}

File: dotnet/src/Planners/Planners.Core/Action/ActionPlanner.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;
using Microsoft.SemanticKernel.AI;
using Microsoft.SemanticKernel.Planning.Action;

namespace Microsoft.SemanticKernel.Planning;

/// <summary>
/// Action Planner allows to select one function out of many, to achieve a given goal.
/// The planner implement the Intent Detection pattern, uses the functions registered
/// in the kernel to see if there's a relevant one, providing instructions to call the
/// function and the rationale used to select it. The planner can also return
/// "no function" is nothing relevant is available.
/// The rationale is currently available only in the prompt, we might include it in
/// the Plan object in future.
/// </summary>
public sealed class ActionPlanner
{
    private const string StopSequence = "#END-OF-PLAN";
    private const string PluginName = "this";

    /// <summary>
    /// The regular expression for extracting serialized plan.
    /// </summary>
    private static readonly Regex s_planRegex = new("^[^{}]*(((?'Open'{)[^{}]*)+((?'Close-Open'})[^{}]*)+)*(?(Open)(?!))", RegexOptions.Singleline | RegexOptions.Compiled);

    /// <summary>Deserialization options for use with <see cref="ActionPlanResponse"/>.</summary>
    private static readonly JsonSerializerOptions s_actionPlayResponseOptions = new()
    {
        AllowTrailingCommas = true,
        DictionaryKeyPolicy = null,
        DefaultIgnoreCondition = JsonIgnoreCondition.Never,
        PropertyNameCaseInsensitive = true,
    };

    // Planner semantic function
    private readonly KernelFunction _plannerFunction;

    private readonly ContextVariables _contextVariables;
    private readonly Kernel _kernel;
    private readonly ILogger _logger;

    // TODO: allow to inject plugin store
    /// <summary>
    /// Initialize a new instance of the <see cref="ActionPlanner"/> class.
    /// </summary>
    /// <param name="kernel">The semantic kernel instance.</param>
    /// <param name="config">The planner configuration.</param>
    public ActionPlanner(
        Kernel kernel,
        ActionPlannerConfig? config = null)
    {
        Verify.NotNull(kernel);
        this._kernel = kernel;

        // Set up Config with default values and excluded plugins
        this.Config = config ?? new();
        this.Config.ExcludedPlugins.Add(PluginName);

        string promptTemplate = this.Config.GetPromptTemplate?.Invoke() ?? EmbeddedResource.Read("Action.skprompt.txt");

        this._plannerFunction = kernel.CreateFunctionFromPrompt(
            promptTemplate: promptTemplate,
            new PromptExecutionSettings()
            {
                ExtensionData = new()
                {
                    { "StopSequences", new[] { StopSequence } },
                    { "MaxTokens", this.Config.MaxTokens },
                }
            });

        kernel.ImportPluginFromObject(this, pluginName: PluginName);

        // Create context and logger
        this._contextVariables = new ContextVariables();
        this._logger = kernel.LoggerFactory.CreateLogger(this.GetType()) ?? NullLogger.Instance;
    }

    /// <summary>Creates a plan for the specified goal.</summary>
    /// <param name="goal">The goal for which a plan should be created.</param>
    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
    /// <returns>The created plan.</returns>
    /// <exception cref="ArgumentNullException"><paramref name="goal"/> is null.</exception>
    /// <exception cref="ArgumentException"><paramref name="goal"/> is empty or entirely composed of whitespace.</exception>
    /// <exception cref="KernelException">A plan could not be created.</exception>
    public Task<Plan> CreatePlanAsync(string goal, CancellationToken cancellationToken = default)
    {
        Verify.NotNullOrWhiteSpace(goal);

        return PlannerInstrumentation.CreatePlanAsync(
            static (ActionPlanner planner, string goal, CancellationToken cancellationToken) => planner.CreatePlanCoreAsync(goal, cancellationToken),
            static (Plan plan) => plan.ToSafePlanString(),
            this, goal, this._logger, cancellationToken);
    }

    private async Task<Plan> CreatePlanCoreAsync(string goal, CancellationToken cancellationToken)
    {
        this._contextVariables.Update(goal);

        FunctionResult result = await this._plannerFunction.InvokeAsync(this._kernel, this._contextVariables, cancellationToken: cancellationToken).ConfigureAwait(false);
        ActionPlanResponse? planData = this.ParsePlannerResult(result);

        if (planData == null)
        {
            throw new KernelException("The plan deserialized to a null object");
        }

        // Build and return plan
        Plan? plan = null;

        FunctionUtils.SplitPluginFunctionName(planData.Plan.Function, out var pluginName, out var functionName);
        if (!string.IsNullOrEmpty(functionName))
        {
            var getFunctionCallback = this.Config.GetFunctionCallback ?? this._kernel.Plugins.GetFunctionCallback();
            var pluginFunction = getFunctionCallback(pluginName, functionName);
            if (pluginFunction != null)
            {
                plan = new Plan(goal, pluginFunction);
                plan.Steps[0].PluginName = pluginName;
            }
        }

        plan ??= new(goal);

        // Populate plan parameters using the function and the parameters suggested by the planner
        if (plan.Steps.Count > 0)
        {
            foreach (KeyValuePair<string, object> p in planData.Plan.Parameters)
            {
                if (p.Value?.ToString() is string value)
                {
                    plan.Steps[0].Parameters[p.Key] = value;
                }
            }
        }

        return plan;
    }

    // TODO: use goal to find relevant functions in a plugin store
    /// <summary>
    /// Native function returning a list of all the functions in the current context,
    /// excluding functions in the planner itself.
    /// </summary>
    /// <param name="goal">Currently unused. Will be used to handle long lists of functions.</param>
    /// <param name="cancellationToken">The token to use to request cancellation.</param>
    /// <returns>List of functions, formatted accordingly to the prompt</returns>
    [KernelFunction, Description("List all functions available in the kernel")]
    public async Task<string> ListOfFunctionsAsync(
        [Description("The current goal processed by the planner")] string goal,
        CancellationToken cancellationToken = default)
    {
        // Prepare list using the format used by skprompt.txt
        var list = new StringBuilder();
        var availableFunctions = await this._kernel.Plugins.GetFunctionsAsync(this.Config, goal, this._logger, cancellationToken).ConfigureAwait(false);
        this.PopulateList(list, availableFunctions);

        return list.ToString();
    }

    // TODO: generate string programmatically
    // TODO: use goal to find relevant examples
    /// <summary>
    /// Native function that provides a list of good examples of plans to generate.
    /// </summary>
    /// <param name="goal">The current goal processed by the planner.</param>
    /// <param name="variables">Function execution context variables.</param>
    /// <returns>List of good examples, formatted accordingly to the prompt.</returns>
    [KernelFunction, Description("List a few good examples of plans to generate")]
    public string GoodExamples(
        [Description("The current goal processed by the planner")] string goal,
        ContextVariables variables)
    {
        return @"
[EXAMPLE]
- List of functions:
// Read a file.
FileIOPlugin.ReadAsync
Parameter ""path"": Source file.
// Write a file.
FileIOPlugin.WriteAsync
Parameter ""path"": Destination file. (default value: sample.txt)
Parameter ""content"": File content.
// Get the current time.
TimePlugin.Time
No parameters.
// Makes a POST request to a uri.
HttpPlugin.PostAsync
Parameter ""body"": The body of the request.
- End list of functions.
Goal: create a file called ""something.txt"".
{""plan"":{
""rationale"": ""the list contains a function that allows to create files"",
""function"": ""FileIOPlugin.WriteAsync"",
""parameters"": {
""path"": ""something.txt"",
""content"": null
}}}
#END-OF-PLAN
";
    }

    // TODO: generate string programmatically
    /// <summary>
    /// Native function that provides a list of edge case examples of plans to handle.
    /// </summary>
    /// <param name="goal">The current goal processed by the planner.</param>
    /// <param name="variables">Function execution context variables.</param>
    /// <returns>List of edge case examples, formatted accordingly to the prompt.</returns>
    [KernelFunction, Description("List a few edge case examples of plans to handle")]
    public string EdgeCaseExamples(
        [Description("The current goal processed by the planner")] string goal,
        ContextVariables variables)
    {
        return @"
[EXAMPLE]
- List of functions:
// Get the current time.
TimePlugin.Time
No parameters.
// Write a file.
FileIOPlugin.WriteAsync
Parameter ""path"": Destination file. (default value: sample.txt)
Parameter ""content"": File content.
// Makes a POST request to a uri.
HttpPlugin.PostAsync
Parameter ""body"": The body of the request.
// Read a file.
FileIOPlugin.ReadAsync
Parameter ""path"": Source file.
- End list of functions.
Goal: tell me a joke.
{""plan"":{
""rationale"": ""the list does not contain functions to tell jokes or something funny"",
""function"": """",
""parameters"": {
}}}
#END-OF-PLAN
";
    }

    #region private ================================================================================

    /// <summary>
    /// The configuration for the ActionPlanner
    /// </summary>
    private ActionPlannerConfig Config { get; }

    /// <summary>
    /// Native function that filters out good JSON from planner result in case additional text is present
    /// using a similar regex to the balancing group regex defined here: https://learn.microsoft.com/en-us/dotnet/standard/base-types/grouping-constructs-in-regular-expressions#balancing-group-definitions
    /// </summary>
    /// <param name="plannerResult">Result of planner function.</param>
    /// <returns>Instance of <see cref="ActionPlanResponse"/> object deserialized from extracted JSON.</returns>
    private ActionPlanResponse? ParsePlannerResult(FunctionResult plannerResult)
    {
        if (plannerResult.GetValue<string>() is string result)
        {
            Match match = s_planRegex.Match(result);

            if (match.Success && match.Groups["Close"] is { Length: > 0 } close)
            {
                string planJson = $"{{{close}}}";
                try
                {
                    return JsonSerializer.Deserialize<ActionPlanResponse?>(planJson, s_actionPlayResponseOptions);
                }
                catch (Exception e)
                {
                    throw new KernelException("Plan parsing error, invalid JSON", e);
                }
            }
        }

        throw new KernelException($"Failed to extract valid json string from planner result: '{plannerResult}'");
    }

    private void PopulateList(StringBuilder list, IEnumerable<KernelFunctionMetadata> functions)
    {
        foreach (KernelFunctionMetadata func in functions)
        {
            // Function description
            if (func.Description != null)
            {
                list.AppendLine($"// {AddPeriod(func.Description)}");
            }
            else
            {
                this._logger.LogWarning("{0}.{1} is missing a description", func.PluginName, func.Name);
                list.AppendLine($"// Function {func.PluginName}.{func.Name}.");
            }

            // Function name
            list.AppendLine($"{func.PluginName}.{func.Name}");

            // Function parameters
            foreach (var p in func.Parameters)
            {
                var description = string.IsNullOrEmpty(p.Description) ? p.Name : p.Description!;
                var defaultValueString = string.IsNullOrEmpty(p.DefaultValue) ? string.Empty : $" (default value: {p.DefaultValue})";
                list.AppendLine($"Parameter \"{p.Name}\": {AddPeriod(description)} {defaultValueString}");
            }
        }
    }

    private static string AddPeriod(string x)
    {
        return x.EndsWith(".", StringComparison.Ordinal) ? x : $"{x}.";
    }

    #endregion
}
